File: frontend/src/utils/ratingCalc.js
/**
 * Расчёт рейтинга по формулам ФНТР (дублирует бота для превью).
 * КС: 3:0/0:3 → 1.2, 3:1/1:3 → 1.0, 3:2/2:3 → 0.8
 * ПРв = (100 – (РТВ – РТП)) / 10 * КД * КС
 * ПРп = -(100 – (РТВ – РТП)) / 20 * КД * КС
 */

export function calculateScoreCoef(setsP1, setsP2) {
  const a = Number(setsP1)
  const b = Number(setsP2)
  if ((a === 3 && b === 0) || (a === 0 && b === 3)) return 1.2
  if ((a === 3 && b === 1) || (a === 1 && b === 3)) return 1.0
  if ((a === 3 && b === 2) || (a === 2 && b === 3)) return 0.8
  return 1.0
}

export function calculateMatchRating(winnerRating, loserRating, winnerSets, loserSets, kd) {
  const ks = calculateScoreCoef(winnerSets, loserSets)
  const base = (100 - (winnerRating - loserRating)) / 10
  const deltaWinner = Math.round(base * kd * ks * 100) / 100
  const deltaLoser = Math.round(-(base / 2) * kd * ks * 100) / 100
  return { deltaWinner, deltaLoser }
}

export function calcRatingDelta(ratingWinner, ratingLoser, divisionCoef, setsWinner, setsLoser) {
  return calculateMatchRating(ratingWinner, ratingLoser, setsWinner, setsLoser, divisionCoef)
}

File: api/rating_calc.py
"""
Rating delta by ФНТР formulas. Mirrors bot/services/rating_calculator.py.
"""
from decimal import Decimal
from typing import Tuple


def _score_coef(sets_winner: int, sets_loser: int) -> float:
    if (sets_winner == 3 and sets_loser == 0) or (sets_winner == 0 and sets_loser == 3):
        return 1.2
    if (sets_winner == 3 and sets_loser == 1) or (sets_winner == 1 and sets_loser == 3):
        return 1.0
    if (sets_winner == 3 and sets_loser == 2) or (sets_winner == 2 and sets_loser == 3):
        return 0.8
    return 1.0


def calculate_match_rating(
    winner_rating: float,
    loser_rating: float,
    winner_sets: int,
    loser_sets: int,
    kd: float,
) -> Tuple[float, float]:
    if winner_sets == 0 and loser_sets == 0:
        return (0.0, 0.0)
    rw = Decimal(str(winner_rating))
    rl = Decimal(str(loser_rating))
    kd_dec = Decimal(str(kd))
    ks = Decimal(str(_score_coef(winner_sets, loser_sets)))
    diff = rw - rl
    base = (Decimal("100") - diff) / Decimal("10")
    delta_winner = float((base * kd_dec * ks).quantize(Decimal("0.01")))
    delta_loser = float(-(base / Decimal("2") * kd_dec * ks).quantize(Decimal("0.01")))
    return (delta_winner, delta_loser)
